sklearn/metrics/pairwise.py

```diff
@@ -224,41 +240,157 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
     """
     X, Y = check_pairwise_arrays(X, Y)
 
-    if X_norm_squared is not None:
-        XX = check_array(X_norm_squared)
-        if XX.shape == (1, X.shape[0]):
-            XX = XX.T
-        elif XX.shape != (X.shape[0], 1):
-            raise ValueError(
-                "Incompatible dimensions for X and X_norm_squared")
-    else:
-        XX = row_norms(X, squared=True)[:, np.newaxis]
+    XX, YY = _check_norms(X, Y, X_norm_squared, Y_norm_squared)
+
+    n_features = X.shape[1]
+
+    # For n_features > 32 we use the 'fast 'method to compute the euclidean
+    # distance, i.e. d(x,y)² = ||x||² + ||y||² - 2 * x.y
+    # It's faster but less precise.
+    if n_features > 32:
```

Comment:
This constant appear twice in the code. Maybe it should be set as a variable so that it could be easily changed in the future.
BTW, I replicated your benchmark `cdist` vs. `sklearn` https://github.com/scikit-learn/scikit-learn/issues/9354#issuecomment-471812174, and it would rather suggest a bound of 8 on my machine.

Reply:
I went half way. 16 seems to be a good compromise. Compared to my first benchmarks, it's not in the much slower than cdist part and compared to yours, it's not in the much slower than master part.